File: src/com/morph/engine/newgui/GUI.java
package com.morph.engine.newgui;

import com.morph.engine.core.Game;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Fernando on 3/6/2017.
 */
public abstract class GUI {
    private List<Element> elements;
    private Game game;

    public GUI(Game game) {
        elements = new ArrayList<>();
        this.game = game;
    }

    public abstract void load();
    public abstract void unload();

    public void addElement(Element e) {
        elements.add(e);
        if (e instanceof Container) {
            ((Container)e).getChildren(true).forEach(this::addElement);
        }
        game.addElement(e);
    }

    public void removeElement(Element e) {
        elements.remove(e);
        if (e instanceof Container) {
            ((Container)e).getChildren(true).forEach(this::removeElement);
        }
        game.removeElement(e);
    }

    public List<Element> getElements() {
        return elements;
    }
}

File: src/main/kotlin/com/morph/engine/newgui/GUI.kt
package com.morph.engine.newgui

import com.morph.engine.core.Game
import java.util.*

/**
 * Created by Fernando on 3/6/2017.
 */
abstract class GUI(val game: Game) {
    private val elements: MutableList<Element>
    var isOpen = false
        private set

    init {
        elements = ArrayList()
    }

    abstract fun init()
    abstract fun load()
    abstract fun unload()

    fun addElement(e: Element) {
        elements.add(e)
        if (e is Container) {
            e.getChildren(true).forEach{ this.addElement(it) }
        }
    }

    fun removeElement(e: Element) {
        elements.remove(e)
        if (e is Container) {
            e.getChildren(true).forEach { this.removeElement(it) }
        }
    }

    fun getElementByName(name: String): Element? {
        return elements.find { it.name == name }
    }

    open fun preUpdate() {}
    open fun update() {}
    open fun fixedUpdate(dt: Float) {}
    open fun postUpdate() {}

    fun getElements(): List<Element> {
        return elements
    }

    fun open() {
        isOpen = true
    }

    fun close() {
        isOpen = false
    }
}
